### PR TITLE
Transaction events parser

### PR DESCRIPTION
**Pull request by popenta**

### CODE REVIEW

File: multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser_test.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    events = find_events_by_first_topic(transaction_outcome, "startPerformAction")
+    parsed = parser.parse_events(events)
+    data = parsed[0]["data"]
+
+    assert data == SimpleNamespace(
+        action_id=1,
+        group_id=0,
+        action_data=SimpleNamespace(
+            **{
```
Comment (andreibancioiu) [inline multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser_test.py:99]: :rocket:

File: multiversx_sdk/abi/abi.py [hunk 2]
```diff
@@ -67,6 +69,15 @@ def __init__(self, definition: AbiDefinition) -> None:
 
             self.endpoints_prototypes_by_name[endpoint.name] = endpoint_prototype
 
+        for event in definition.events:
+            input_prototype = self._create_event_input_prototypes(event)
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/abi/abi.py:73]: removed

File: multiversx_sdk/abi/abi.py [hunk 3]
```diff
@@ -126,10 +137,23 @@ def _create_endpoint_output_prototypes(self, endpoint: EndpointDefinition) -> Li
...
+            parameter_prototype = EventField(name=topic.name, value=self._create_topic_prototype(topic))
+            prototypes.append(parameter_prototype)
+
+        return prototypes
+
     def _create_parameter_prototype(self, parameter: ParameterDefinition) -> Any:
         type_formula = self._type_formula_parser.parse_expression(parameter.type)
         return self._create_prototype(type_formula)
 
+    def _create_topic_prototype(self, parameter: EventTopicDefinition) -> Any:
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/abi/abi.py:153]: Renamed

File: multiversx_sdk/abi/abi.py [hunk 4]
```diff
@@ -126,10 +137,23 @@ def _create_endpoint_output_prototypes(self, endpoint: EndpointDefinition) -> Li
 
         return prototypes
 
+    def _create_event_input_prototypes(self, event: EventDefinition) -> List[Any]:
+        prototypes: List[Any] = []
+
+        for topic in event.inputs:
+            parameter_prototype = EventField(name=topic.name, value=self._create_topic_prototype(topic))
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/abi/abi.py:144]: renamed to `event_field_prototype`.

File: multiversx_sdk/abi/abi.py [hunk 5]
```diff
@@ -248,3 +323,14 @@ class EndpointPrototype:
...
+
+
+class EventField:
+    def __init__(self, name: str, value: Any) -> None:
+        self.name = name
+        self.value = value
+
+
+class EventPrototype:
+    def __init__(self, input_parameters: List[EventField]) -> None:
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/abi/abi.py:335]: renamed

File: multiversx_sdk/abi/abi.py [hunk 6]
```diff
@@ -163,6 +187,49 @@ def decode_endpoint_output_parameters(self, endpoint_name: str, encoded_values:
         output_native_values = [value.get_payload() for value in output_values_as_native_object_holders]
         return output_native_values
 
+    def decode_event(self, event_definition: EventDefinition, topics: List[bytes], data_items: List[bytes]) -> List[Any]:
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/abi/abi.py:190]: Done

File: multiversx_sdk/abi/abi.py [hunk 7]
```diff
@@ -163,6 +187,49 @@ def decode_endpoint_output_parameters(self, endpoint_name: str, encoded_values:
...
+
+        output_values_as_native_object_holders = cast(List[IPayloadHolder], output_values)
+        output_native_values = [value.get_payload() for value in output_values_as_native_object_holders]
+
+        for i in range(len(non_indexed_inputs)):
+            result[non_indexed_inputs[i].name] = output_native_values[i]
+
+        return result
+
+    def get_event(self, name: str) -> EventDefinition:
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/abi/abi.py:222]: moved the method in `AbiDefinition`.

File: multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        return [self.parse_event(event) for event in events]
+
+    def parse_event(self, event: TransactionEvent) -> Any:
+        first_topic = event.topics[0].decode() if len(event.topics) else ""
+        abi_identifier = first_topic if self.first_topic_as_identifier else event.identifier
+
+        if self.first_topic_as_identifier:
+            event.topics = event.topics[1:]
+
+        event_definition = self.abi.get_event(abi_identifier)
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py:36]: done

File: multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    def parse_events(self, events: List[TransactionEvent]) -> List[Any]:
+        return [self.parse_event(event) for event in events]
+
+    def parse_event(self, event: TransactionEvent) -> Any:
+        first_topic = event.topics[0].decode() if len(event.topics) else ""
+        abi_identifier = first_topic if self.first_topic_as_identifier else event.identifier
+
+        if self.first_topic_as_identifier:
+            event.topics = event.topics[1:]
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py:34]: now using `deepcopy` to copy `event.topics` to another `list` and I am modifying that list.

File: multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py [hunk 10]
```diff
@@ -23,19 +21,20 @@ def __init__(self, abi: IAbi, first_topic_as_identifier: bool = True) -> None:
...
-    def parse_events(self, events: List[TransactionEvent]) -> List[Any]:
+    def parse_events(self, events: List[TransactionEvent]) -> List[SimpleNamespace]:
         return [self.parse_event(event) for event in events]
 
-    def parse_event(self, event: TransactionEvent) -> Any:
+    def parse_event(self, event: TransactionEvent) -> SimpleNamespace:
         first_topic = event.topics[0].decode() if len(event.topics) else ""
         abi_identifier = first_topic if self.first_topic_as_identifier else event.identifier
 
+        topics = deepcopy(event.topics)
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py:31]: indeed, you are right

File: multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser_test.py [hunk 11]
```diff
@@ -27,7 +27,7 @@ def test_parse_events_minimalistic():
     )
 
     assert len(values) == 1
-    assert values[0] == {"batch_id": 42, "tx_id": 43}
+    assert values[0] == SimpleNamespace(**{"batch_id": 42, "tx_id": 43})
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser_test.py:30]: indeed, fixed here and bellow

**Review by andreibancioiu — APPROVED** [review]

File: multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        # This is true for log entries emitted by smart contracts:
+        # https://github.com/multiversx/mx-chain-vm-go/blob/v1.5.27/vmhost/contexts/output.go#L270
+        # https://github.com/multiversx/mx-chain-vm-go/blob/v1.5.27/vmhost/contexts/output.go#L283
+        self.first_topic_as_identifier = first_topic_as_identifier
+
+    def parse_events(self, events: List[TransactionEvent]) -> List[SimpleNamespace]:
+        return [self.parse_event(event) for event in events]
+
+    def parse_event(self, event: TransactionEvent) -> SimpleNamespace:
+        first_topic = event.topics[0].decode() if len(event.topics) else ""
```
Comment (axenteoctavian) [inline multiversx_sdk/core/transactions_outcome_parsers/transaction_events_parser.py:27]: I'm not sure if it's possible but...
If the event has 0 topics and you consider first topic as identifier, this function returns something bad. Maybe you can consider `event.identifier` as event ID `if first_topic == "" && self.first_topic_as_identifier`

File: multiversx_sdk/testutils/testdata/esdt-safe.abi.json
Comment (axenteoctavian) [inline multiversx_sdk/testutils/testdata/esdt-safe.abi.json:1]: FYI
This version of abi is very old but it's fine for your tests. :smile:

Comment (popenta) [reply multiversx_sdk/testutils/testdata/esdt-safe.abi.json:1]: maybe we'll replace it in the future

**Review by andreibancioiu — APPROVED** [review]

**Review by axenteoctavian — APPROVED** [review]

**Pull request merged**